### ISSUE TITLE
Problem with apache removing the gd module

### ISSUE THREAD

**Issue by SgiarC** [phpbrew/phpbrew#1256]
There is a line in the README.md that states:
"Compile apache php module and separate them by different versions."

I can't find the instructions to create an apache module something like:
mod_php-5.6.40.so

On Fedora 34
I can get phpbrew to create php-5.6.40 with gd and it shows up using ./php or ./php-cgi but when apache serves the php the gd module is missing. The build dates from the command line or through Apache are the same so I know Apache is pointing to the correct build.

I'm using fcgid to run the php.
#!/bin/sh
PHPRC="/root/.phpbrew/php/php-5.6.40/etc/php.ini"
export PHPRC
export PHP_FCGI_MAX_REQUESTS=3000
export PHP_FCGI_CHILDREN=8
exec /usr/local/php-5.6.40/bin/php-cgi
#

I tried moving php into another directory outside of /root
the original location also does not work
exec /root/.phpbrew/php/php-5.6.40/bin/php-cgi

I have also tried this method:
<FilesMatch \.php$>
    SetHandler "proxy:unix:/run/php-fpm/www.sock|fcgi://127.0.0.1"
</FilesMatch>
This will load the current php version, so I know the "proxy:unix" method works. The user and group of www.sock are root:root

<FilesMatch \.php$>
   SetHandler "proxy:unix:/root/.phpbrew/php/php-5.6.40/var/run/php-fpm.sock|fcgi://127.0.0.1"
</FilesMatch>
If I try to load the php-5.6.40 version using 'proxy:unix:', it gives a

Service Unavailable
The server is temporarily unable to service your request due to maintenance downtime or capacity problems. Please try again later.

and in the logs it states: 
FCGI: attempt to connect to Unix domain socket /root/.phpbrew/php/php-5.6.40/var/run/php-fpm.sock (*) failed
 failed to make connection to backend: httpd-UDS
The user and group of php-fpm.sock are both apache 660, root:root also does not work.

Here is the list of get_loaded_extensions() that apache serves up. [43] should be gd but it gets switched to cgi-fcgi for some reason.
    [7] => bcmath
    [8] => bz2
    [9] => calendar
    [43] => cgi-fcgi  *********
    [0] => Core
    [10] => ctype
    [11] => curl
    [1] => date
    [12] => dom
    [2] => ereg
    [13] => fileinfo
    [14] => filter
    [15] => hash
    [16] => json
    [3] => libxml
    [17] => mbstring
    [44] => mhash
    [33] => mysql
    [29] => mysqli
    [22] => mysqlnd
    [4] => openssl
    [21] => pcntl
    [5] => pcre
    [23] => PDO
    [24] => pdo_mysql
    [25] => Phar
    [26] => posix
    [27] => readline
    [28] => Reflection
    [19] => session
    [30] => shmop
    [31] => SimpleXML
    [32] => sockets
    [18] => SPL
    [20] => standard
    [34] => sysvmsg
    [35] => sysvsem
    [36] => sysvshm
    [37] => tokenizer
    [38] => xml
    [39] => xmlreader
    [40] => xmlwriter
    [41] => xsl
    [42] => zip
    [6] => zlib

Here is the list of get_loaded_extensions() from the command line ./php or ./php-cgi

    [7] => bcmath
    [8] => bz2
    [9] => calendar
    [0] => Core
    [10] => ctype
    [11] => curl
    [1] => date
    [12] => dom
    [2] => ereg
    [13] => fileinfo
    [14] => filter
    [43] => gd ************
    [15] => hash
    [16] => json
    [3] => libxml
    [17] => mbstring
    [44] => mhash
    [33] => mysql
    [29] => mysqli
    [22] => mysqlnd
    [4] => openssl
    [21] => pcntl
    [5] => pcre
    [23] => PDO
    [24] => pdo_mysql
    [25] => Phar
    [26] => posix
    [27] => readline
    [28] => Reflection
    [19] => session
    [30] => shmop
    [31] => SimpleXML
    [32] => sockets
    [18] => SPL
    [20] => standard
    [34] => sysvmsg
    [35] => sysvsem
    [36] => sysvshm
    [37] => tokenizer
    [38] => xml
    [39] => xmlreader
    [40] => xmlwriter
    [41] => xsl
    [42] => zip
    [6] => zlib

So basically I seem stuck. Any help pointing me in the right direction would be appreciated.
Thanks,
Craig

**Comment by SgiarC** [comment]
I had most of these installed but it did pick up a few of them.

yum install make gcc gcc-c++ kernel-devel # these are equivalent to Debian's "build-essential"
yum install php php-bz2 php-devel php-pear bzip2-devel yum-utils libmcrypt-devel libpqxx-devel libxslt-devel pcre-devel libcurl-devel libgsasl-devel openldap-devel
yum-builddep php

Then I compiled with +apxs2

That created the libphp5.6.40.so module file.
I created a file in the config.modules.d directory that loads the module.
LoadModule php5_module  modules/libphp5.6.40.so

Then in httpd.conf under the virtual host I used.
<FilesMatch \.php$>
	AddType application/x-httpd-php .php
	SetHandler application/x-httpd-php
</FilesMatch>

I'm not sure of the best way to SetHandler but this seems to load the correct php version for that domain.
Now that I look at it the AddType seems redundant inside the FilesMatch.

Inside a virtualHost, what is the best way to use that libphp5.6.40.so that phpbrew created?

Any way that still didn't solve the problem of the gd module getting dropped by apache.

Any ideas?
Thanks, Craig

**Comment by SgiarC** [comment]
Here is something that I noticed in the build log. 
Checking patch for replace freetype-config with pkg-config on php older than 7.4
---> Applying patch...
patching file ext/gd/config.m4
Reversed (or previously applied) patch detected!  Skipping patch.
patch unexpectedly ends in middle of line
2 out of 2 hunks ignored -- saving rejects to file ext/gd/config.m4.rej

Patch failed
0 changes patched.

So I went to the exe/gd folder and did a
make test
It completely failed

Then I loaded the php version using
phpbrew use php-5.6.40

Then I rebuilt the exe/gd using

`phpbrew ext install gd --  --enable-gd-native-ttf  --with-gd  --with-freetype-dir=no  --with-jpeg-dir  --with-libdir  --with-png-dir  --with-vpx-dir  --with-xpm-dir --with-Xpm-dir --with-t1lib

After that in exe/gd make test was 95% successful, it just skipped about 5 tests.

Then I rebuilt using `

`phpbrew  install 5.6.40 +default +mysql +fpm +gd +opcache +apxs2 +openssl=/usr --  \`

This builds, but will never install, it hangs at the install point.
Going into the build directory and running make test it mostly passes, but trying to manually make install it fails, but at least it has built the .so module and placed it in the apache modules symbolic linked directory.

Then I rebuilt once more but without the +apxs2

This one builds and installs.
Then starting apache again and looking at a couple of web pages that access php differently it works both ways CGI/FastCGI and Apache 2.0 Handler,

--------
Server API | CGI/FastCGI

GD Support | enabled

GD Version | bundled (2.1.0 compatible)
FreeType Support | enabled
FreeType Linkage | with freetype
FreeType Version | 2.10.4
GIF Read Support | enabled
GIF Create Support | enabled
JPEG Support | enabled
libJPEG Version | 6b
PNG Support | enabled
libPNG Version | 1.6.37
WBMP Support | enabled
XBM Support | enabled

And the other method of loading php which is using the .so module that phpbrew built:
--------
Server API | Apache 2.0 Handler

GD Support | enabled

GD Version | bundled (2.1.0 compatible)
FreeType Support | enabled
FreeType Linkage | with freetype
FreeType Version | 2.10.4
GIF Read Support | enabled
GIF Create Support | enabled
JPEG Support | enabled
libJPEG Version | 6b
PNG Support | enabled
libPNG Version | 1.6.37
WBMP Support | enabled
XBM Support | enabled

Next I'm going to try to get apache to load php using the
SetHandler "proxy:unix:
method, without giving the "Service Unavailable" error.
